pkg/k8s/custom_source.go

```diff
@@ -0,0 +1,174 @@
+
+	// stop is to end ongoing goroutine, and close the Create channel
+	stop <-chan struct{}
+
+	Source <-chan GenericEvent
+
+	// dest is the destination channels of the Pod event handlers
+	dest []chan GenericEvent
+
+	// DestBufferSize is the specified buffer size of dest channels.
+	// Default to 1024 if not specified.
+	DestBufferSize int
+
+	// destLock is to ensure the destination channels are safely added/removed
+	destLock sync.Mutex
+}
+
+var _ inject.Stoppable = &NotificationChannel{}
+
+// InjectStopChannel is internal should be called only by the Controller.
+// It is used to inject the stop channel initialized by the ControllerManager.
+func (cs *NotificationChannel) InjectStopChannel(stop <-chan struct{}) error {
+	if cs.stop == nil {
+		cs.stop = stop
+	}
+
+	return nil
+}
+
+func (cs *NotificationChannel) String() string {
+	return fmt.Sprintf("channel source: %p", cs)
+}
+
+// Start implements Source and should only be called by the Controller.
+func (cs *NotificationChannel) Start(
+	handler handler.EventHandler,
+	queue workqueue.RateLimitingInterface,
+	prct ...predicate.Predicate) error {
+	// Source should have been specified by the user.
+	if cs.Source == nil {
+		return fmt.Errorf("must specify CustomChannle.Source")
+	}
+
+	// stop should have been injected before Start was called
+	if cs.stop == nil {
+		return fmt.Errorf("must call InjectStop on Channel before calling Start")
+	}
+
+	// use default value if DestBufferSize not specified
+	if cs.DestBufferSize == 0 {
+		cs.DestBufferSize = defaultBufferSize
+	}
+
+	cs.once.Do(func() {
+		// Distribute GenericEvents to all EventHandler / Queue pairs Watching this source
+		go cs.syncLoop()
+	})
+
+	dst := make(chan GenericEvent, cs.DestBufferSize)
+	go func() {
+		for evt := range dst {
+			switch evt.EventType {
+			case CREATE:
+				handler.Create(event.CreateEvent{Meta: evt.Meta, Object: evt.Object}, queue)
+			case DELETE:
+				handler.Delete(event.DeleteEvent{Meta: evt.OldMeta, Object: evt.OldObject}, queue)
+			case UPDATE:
+				handler.Update(event.UpdateEvent{MetaOld: evt.OldMeta, ObjectOld: evt.OldObject, MetaNew: evt.Meta, ObjectNew: evt.Object}, queue)
+			default:
+				_ = fmt.Errorf("Invalid Type %T", evt.EventType)
+			}
+		}
+	}()
+
+	cs.destLock.Lock()
+	defer cs.destLock.Unlock()
+
+	cs.dest = append(cs.dest, dst)
+
+	return nil
+}
+
+func (cs *NotificationChannel) doStop() {
+	cs.destLock.Lock()
+	defer cs.destLock.Unlock()
+
+	for _, dst := range cs.dest {
+		close(dst)
+	}
+}
+
+func (cs *NotificationChannel) distribute(evt GenericEvent) {
+	cs.destLock.Lock()
+	defer cs.destLock.Unlock()
+
+	for _, dst := range cs.dest {
+		// We cannot make it under goroutine here, or we'll meet the
+		// race condition of writing message to closed channels.
+		// To avoid blocking, the dest channels are expected to be of
+		// proper buffer size. If we still see it blocked, then
+		// the controller is thought to be in an abnormal state.
+		dst <- evt
+	}
+}
+
```

Comment:
Can we add description of each function in comments so that it's easy to understand what each function is doing.